### ISSUE TITLE
PL_keyword_plugin needs thread-safe wrap setter

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#16229]
Migrated from [rt.perl.org#132413](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=132413) (status was 'open')

Searchable as RT132413$

**Comment by p5pRT** [comment]
### From @mauke

#### Created by @mauke



PL\_keyword\_plugin \(see perldoc perlapi\) is global to the entire process and
shared across threads\. Its documentation says&#8203;:

| \[\.\.\.\] take a copy of PL\_keyword\_plugin before assigning your own function
| pointer to it\. Your handler function should look for keywords that it is
| interested in and handle those\. Where it is not interested\, it should call
| the saved plugin function\, passing on the arguments it received\.

But it does not show any example code of how this "taking a copy" is supposed
to be done\.

The XS&#8203;::APItest module in the core test suite does it the following way
\(excerpt\)&#8203;:

| static int \(\*next\_keyword\_plugin\)\(pTHX\_ char \*\, STRLEN\, OP \*\*\);
| 
| static int my\_keyword\_plugin\(pTHX\_
|     char \*keyword\_ptr\, STRLEN keyword\_len\, OP \*\*op\_ptr\)
| \{
|     if \(\.\.\.\) \{
|      \.\.\.
|     \} else \{
|      return next\_keyword\_plugin\(aTHX\_ keyword\_ptr\, keyword\_len\, op\_ptr\);
|     \}
| \}
| 
| BOOT&#8203;:
| \{
|     next\_keyword\_plugin = PL\_keyword\_plugin;
|     PL\_keyword\_plugin = my\_keyword\_plugin;
| \}
 
This code has found its way into various CPAN modules \(e\.g\.
Function&#8203;::Parameters\, Syntax&#8203;::Keyword&#8203;::Try\, etc\)\.

It is also not thread\-safe&#8203;:

% perl \-Mthreads \-e 'threads\->create\(sub \{ require Function&#8203;::Parameters \}\)\->join for 1 \.\. 2'
Segmentation fault

This is because each thread runs BOOT separately\, but PL\_keyword\_plugin is
shared\. This means the second thread BOOTing will set next\_keyword\_plugin to
the value the first thread left in PL\_keyword\_plugin\, which is
my\_keyword\_plugin\.

Then the next call to my\_keyword\_plugin that enters the 'else' block will
attempt to run the next handler in the chain \(via 'return
next\_keyword\_plugin\(\.\.\.\)'\)\, but now next\_keyword\_plugin == my\_keyword\_plugin\.

In other words\, the list of linked handlers contains a loop\, leading to a stack overflow\.


PL\_check \(the other global variable documented in perlapi\) has this paragraph
in its documentation&#8203;:

| For thread safety\, modules should not write directly to this array\. Instead\,
| use the function wrap\_op\_checker\.

This equally applies to PL\_keyword\_plugin\, but a corresponding
wrap\_keyword\_plugin function is missing\. It should be added\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.26.0:

Configured by mauke at Fri Sep 22 13:28:36 CEST 2017.

Summary of my perl5 (revision 5 version 26 subversion 0) configuration:
   
  Platform:
    osname=linux
    osvers=4.9.41-1-lts
    archname=i686-linux
    uname='linux simplicio 4.9.41-1-lts #1 smp mon aug 7 17:57:02 cest 2017 i686 gnulinux '
    config_args=''
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=undef
    use64bitall=undef
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
    bincompat5005=undef
  Compiler:
    cc='cc'
    ccflags ='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    optimize='-O2 -march=native'
    cppflags='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='7.2.0'
    gccosandvers=''
    intsize=4
    longsize=4
    ptrsize=4
    doublesize=8
    byteorder=1234
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=12
    longdblkind=3
    ivtype='long'
    ivsize=4
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=4
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags ='-fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/gcc/i686-pc-linux-gnu/7.2.0/include-fixed /usr/lib /lib
    libs=-lpthread -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.26.so
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version='2.26'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -O2 -march=native -L/usr/local/lib -fstack-protector-strong'



@INC for perl 5.26.0:
    /home/mauke/usr/lib/perl5/site_perl/5.26.0/i686-linux
    /home/mauke/usr/lib/perl5/site_perl/5.26.0
    /home/mauke/usr/lib/perl5/5.26.0/i686-linux
    /home/mauke/usr/lib/perl5/5.26.0


Environment for perl 5.26.0:
    HOME=/home/mauke
    LANG=en_US.UTF-8
    LANGUAGE=en_US
    LC_COLLATE=C
    LC_MONETARY=de_DE.UTF-8
    LC_TIME=de_DE.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/mauke/perl5/perlbrew/bin:/home/mauke/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl
    PERLBREW_BASHRC_VERSION=0.73
    PERLBREW_HOME=/home/mauke/.perlbrew
    PERLBREW_ROOT=/home/mauke/perl5/perlbrew
    PERL_BADLANG (unset)
    PERL_UNICODE=SAL
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @leonerd

On Tue\, 07 Nov 2017 09&#8203;:44&#8203;:49 \-0800
l\.mai@&#8203;web\.de \(via RT\) \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> This code has found its way into various CPAN modules \(e\.g\.
> Function&#8203;::Parameters\, Syntax&#8203;::Keyword&#8203;::Try\, etc\)\.

In fact this issue came to light precisely because someone found it in
S&#8203;:K&#8203;:T&#8203;:

&nbsp;  https://rt.cpan.org/Ticket/Display.html?id=123547

I've applied the same fix that mauke did; namely to borrow the
OP\_CHECK\_MUTEX for also protecting the keyword plugin hook\.

Since that only turned up in perl 5\.14\, I've had to document it as a
known bug\, thus&#8203;:

&nbsp;  https://metacpan.org/pod/release/PEVANS/Syntax-Keyword-Try-0.09/lib/Syntax/Keyword/Try.pm#KNOWN-BUGS

I'd suggest an easy fix would be to either document that this really is
the suggested mechanism other modules use\, or add a new mutex for them
to use instead\.

While this fix now protects S&#8203;:K&#8203;:T against concurrent loading of itself\,
or with Function&#8203;::Parameters\, it won't protect against any other
modules that aren't so locked\. Having a common standard that all
keyword plugin\-using modules use is essential here\.

\-\- 
Paul "LeoNerd" Evans

leonerd@&#8203;leonerd\.org\.uk      |  https://metacpan.org/author/PEVANS
http&#8203;://www\.leonerd\.org\.uk/  |  https://www.tindie.com/stores/leonerd/

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @mauke

On Tue\, 07 Nov 2017 09&#8203;:44&#8203;:49 \-0800\, mauke\- wrote&#8203;:
> 
> PL\_check \(the other global variable documented in perlapi\) has this
> paragraph
> in its documentation&#8203;:
> 
> | For thread safety\, modules should not write directly to this array\.
> Instead\,
> | use the function wrap\_op\_checker\.
> 
> This equally applies to PL\_keyword\_plugin\, but a corresponding
> wrap\_keyword\_plugin function is missing\. It should be added\.

I have done this in a branch&#8203;: https://perl5.git.perl.org/perl.git/shortlog/refs/heads/smoke-me/mauke/keyword-plugin-mutex

If no one screams\, I'll merge this in a couple of days \(if I remember\)\.

**Comment by p5pRT** [comment]
### From @leonerd

On Wed\, 08 Nov 2017 16&#8203;:34&#8203;:25 \-0800
"l\.mai@&#8203;web\.de via RT" \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> I have done this in a branch&#8203;:
> https://perl5.git.perl.org/perl.git/shortlog/refs/heads/smoke-me/mauke/keyword-plugin-mutex
> 
> If no one screams\, I'll merge this in a couple of days \(if I
> remember\)\.

From the perspective of a potential user of this change\, I think that
looks nice\. It neatly wraps up the requirements nicely\.

One question&#8203;: will there be an accompanying update to ppport\.h or a
suggested \#ifndef\-style pattern for back\-compat?

\-\- 
Paul "LeoNerd" Evans

leonerd@&#8203;leonerd\.org\.uk      |  https://metacpan.org/author/PEVANS
http&#8203;://www\.leonerd\.org\.uk/  |  https://www.tindie.com/stores/leonerd/

**Comment by p5pRT** [comment]
### From @xsawyerx



On 11/09/2017 04&#8203;:26 AM\, Paul "LeoNerd" Evans wrote&#8203;:
> On Wed\, 08 Nov 2017 16&#8203;:34&#8203;:25 \-0800
> "l\.mai@&#8203;web\.de via RT" \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>
>> I have done this in a branch&#8203;:
>> https://perl5.git.perl.org/perl.git/shortlog/refs/heads/smoke-me/mauke/keyword-plugin-mutex
>>
>> If no one screams\, I'll merge this in a couple of days \(if I
>> remember\)\.
> From the perspective of a potential user of this change\, I think that
> looks nice\. It neatly wraps up the requirements nicely\.
>
> One question&#8203;: will there be an accompanying update to ppport\.h or a
> suggested \#ifndef\-style pattern for back\-compat?


At the moment there is no active maintainer to Devel&#8203;::PPPort\, so authors
need to add their own backporting abilities\. If it receives updates\, I'm
sure we could make additional releases for it\.

We've also discussed removing support for any Perl under 5\.6\. At the
moment\, Devel&#8203;::PPPort supports 5\.3\.

Lukas\, would you be able to backport this to Devel&#8203;::PPPort?

**Comment by p5pRT** [comment]
### From @mauke

Am 09\.11\.2017 um 10&#8203;:27 schrieb Sawyer X&#8203;:
> On 11/09/2017 04&#8203;:26 AM\, Paul "LeoNerd" Evans wrote&#8203;:
>> On Wed\, 08 Nov 2017 16&#8203;:34&#8203;:25 \-0800
>> "l\.mai@&#8203;web\.de via RT" \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>>
>>> I have done this in a branch&#8203;:
>>> https://perl5.git.perl.org/perl.git/shortlog/refs/heads/smoke-me/mauke/keyword-plugin-mutex
>>>
>>> If no one screams\, I'll merge this in a couple of days \(if I
>>> remember\)\.
>>  From the perspective of a potential user of this change\, I think that
>> looks nice\. It neatly wraps up the requirements nicely\.
>>
>> One question&#8203;: will there be an accompanying update to ppport\.h or a
>> suggested \#ifndef\-style pattern for back\-compat?
> 
> 
> At the moment there is no active maintainer to Devel&#8203;::PPPort\, so authors
> need to add their own backporting abilities\. If it receives updates\, I'm
> sure we could make additional releases for it\.
> 
> We've also discussed removing support for any Perl under 5\.6\. At the
> moment\, Devel&#8203;::PPPort supports 5\.3\.
> 
> Lukas\, would you be able to backport this to Devel&#8203;::PPPort?

I have no idea how Devel&#8203;::PPPort works\.

My suggestion for older perls would be&#8203;:

\#ifndef wrap\_keyword\_plugin
\#define wrap\_keyword\_plugin\(a\,b\) S\_wrap\_keyword\_plugin\(aTHX\_ a\,b\)
static void
S\_wrap\_keyword\_plugin\(pTHX\_ Perl\_keyword\_plugin\_t new\_plugin\, 
Perl\_keyword\_plugin\_t \*old\_plugin\_p\) \{
&nbsp;     dVAR;
&nbsp;     PERL\_UNUSED\_CONTEXT;
&nbsp;     if \(\*old\_plugin\_p\) return;
&nbsp;     OP\_CHECK\_MUTEX\_LOCK;
&nbsp;     if \(\!\*old\_plugin\_p\) \{
&nbsp;         \*old\_plugin\_p = PL\_keyword\_plugin;
&nbsp;         PL\_keyword\_plugin = new\_plugin;
&nbsp;     \}
&nbsp;     OP\_CHECK\_MUTEX\_UNLOCK;
\}
\#endif

Rationale&#8203;:

\- To be useful\, the compat code must use one single shared mutex\. It 
wouldn't help if e\.g\. every XS module created its own mutex\.

\- OP\_CHECK\_MUTEX is technically the wrong mutex \(it protects access to 
PL\_check\)\. But it exists\, is easy to access\, and the only downside is 
that it is slightly overprotective&#8203;: Multi\-threaded updates to 
PL\_keyword\_plugin and PL\_check now wait for each other instead of 
running in parallel\. I think this is harmless because the critical 
sections only consist of two variable assignments; there's nothing that 
could block for a long time\. \(And the situation should be very rare in 
the first place\.\)

\- OP\_CHECK\_MUTEX\_LOCK / OP\_CHECK\_MUTEX\_UNLOCK are not part of the public 
API\. They're not documented anywhere AFAICS\. However\, on the perls where 
they exist \(5\.16 \.\. 5\.26\)\, they are stable and reliable\. We can sanction 
their use retroactively if module authors agree not to use them where 
wrap\_keyword\_plugin is available\. That way we're free to change or 
remove them in the future\.

\- This leaves 5\.12 and 5\.14 out in the cold\. I don't know what to do there\.

Thoughts?

\-\- 
Lukas Mai \<plokinom@&#8203;gmail\.com>

**Comment by p5pRT** [comment]
### From zefram@fysh.org

l\.mai@&#8203;web\.de wrote&#8203;:
>It is also not thread\-safe&#8203;:

There are really two distinct issues around thread safety\, worth
separating out\.

Firstly\, there's the issue you demonstrated\, that the way in which
the former plugin value is saved needs to match the scope of the hook
variable\.  This separates into two aspects in which it needs to match&#8203;:
storage and initialisation\.  Since PL\_keyword\_plugin is once per process\,
a module's saved next\-plugin value also needs to be a one\-per\-process
variable\, and it needs to be initialised once per process\.  Storage is
properly arranged by a "static" C variable\.  For initialisation to be
once per process\, unconditional initialisation in a BOOT section is\,
as you show\, not sufficient\.  Some kind of conditional is required\.

Secondly\, there's the more obscure possibility of a race condition\,
if two hooking operations run simultaneously in different threads\.
To resolve this requires the hooking critical section to be governed by
a mutex\.  In new Perls it could be a mutex dedicated to this purpose\.
In existing Perls it would have to be an existing mutex that we extend
to this purpose; it doesn't matter much which one we use\, but it is
important that we all agree on which one\.

The comparison to wrap\_op\_checker\(\) is apt\.  The scoping is the same\,
and the way the hook variables are used is the same\, so the solution
will be essentially the same\.  Also\, if you look into the history of how
wrap\_op\_checker\(\) came to be\, it arose out of a similar conversation
about thread safety\.  The same issue of agreeing on a mutex to use on
older Perls applies too\.  Whereas wrap\_op\_checker\(\) has its own mutex on
Perls where it's in the core\, there is a de facto agreement that reserve
implementations of it for older Perls will use the OP\_REFCNT mutex\.

I agree that a wrap\_keyword\_plugin\(\) function would be sensible\.
Note that PL\_keyword\_plugin is flagged as experimental\, so
wrap\_keyword\_plugin\(\) should also be so marked\, being part of the same
experiment\.  \(I have things to say about where that experiment should go\,
but that's for another time\.\)  With regard to which mutex to use on older
Perls\, I suggest using the OP\_REFCNT mutex again\.  OP\_CHECK\_MUTEX would
be another option where it exists\, but that's only back to 5\.15\.8\, so
would require reserve definitions to be more complex\, with conditional
code to switch to a different mutex on pre\-5\.15\.8 perls\.  OP\_REFCNT is
available all the way back\.  There's no real performance concern affecting
the choice between these mutexes\, just the maintainability issue\.

\-zefram

**Comment by p5pRT** [comment]
### From zefram@fysh.org

l\.mai@&#8203;web\.de via RT wrote&#8203;:
>I have done this in a branch&#8203;: https://perl5.git.perl.org/perl.git/shortlog/refs/heads/smoke-me/mauke/keyword-plugin-mutex
>
>If no one screams\, I'll merge this in a couple of days \(if I remember\)\.

wrap\_keyword\_plugin\(\) needs to be marked experimental\, to match
PL\_keyword\_plugin\.  Otherwise the patch looks good\.

\-zefram

**Comment by p5pRT** [comment]
### From @mauke

On Thu\, 09 Nov 2017 16&#8203;:24&#8203;:59 \-0800\, zefram@&#8203;fysh\.org wrote&#8203;:
> l\.mai@&#8203;web\.de via RT wrote&#8203;:
> > I have done this in a branch&#8203;:
> > https://perl5.git.perl.org/perl.git/shortlog/refs/heads/smoke-
> > me/mauke/keyword\-plugin\-mutex
> >
> > If no one screams\, I'll merge this in a couple of days \(if I
> > remember\)\.
> 
> wrap\_keyword\_plugin\(\) needs to be marked experimental\, to match
> PL\_keyword\_plugin\.  Otherwise the patch looks good\.

Thanks for the review\. I have pushed a new version of my branch\.

**Comment by p5pRT** [comment]
### From @mauke

On Thu\, 09 Nov 2017 16&#8203;:05&#8203;:54 \-0800\, zefram@&#8203;fysh\.org wrote&#8203;:
> I agree that a wrap\_keyword\_plugin\(\) function would be sensible\.
> Note that PL\_keyword\_plugin is flagged as experimental\, so
> wrap\_keyword\_plugin\(\) should also be so marked\, being part of the same
> experiment\.  \(I have things to say about where that experiment should go\,
> but that's for another time\.\)  With regard to which mutex to use on older
> Perls\, I suggest using the OP\_REFCNT mutex again\.  OP\_CHECK\_MUTEX would
> be another option where it exists\, but that's only back to 5\.15\.8\, so
> would require reserve definitions to be more complex\, with conditional
> code to switch to a different mutex on pre\-5\.15\.8 perls\.  OP\_REFCNT is
> available all the way back\.  There's no real performance concern affecting
> the choice between these mutexes\, just the maintainability issue\.


In light of this\, my back\-compat suggestion becomes&#8203;:

\#ifndef wrap\_keyword\_plugin
\#define wrap\_keyword\_plugin\(a\, b\) S\_wrap\_keyword\_plugin\(aTHX\_ a\, b\)
static void S\_wrap\_keyword\_plugin\(pTHX\_ Perl\_keyword\_plugin\_t new\_plugin\, Perl\_keyword\_plugin\_t \*old\_plugin\_p\) \{
&nbsp;    dVAR;
&nbsp;    PERL\_UNUSED\_CONTEXT;
&nbsp;    if \(\*old\_plugin\_p\) return;
&nbsp;    MUTEX\_LOCK\(&PL\_op\_mutex\);
&nbsp;    if \(\!\*old\_plugin\_p\) \{
&nbsp;        \*old\_plugin\_p = PL\_keyword\_plugin;
&nbsp;        PL\_keyword\_plugin = new\_plugin;
&nbsp;    \}
&nbsp;    MUTEX\_UNLOCK\(&PL\_op\_mutex\);
\}
\#endif

**Comment by p5pRT** [comment]
### From zefram@fysh.org

l\.mai@&#8203;web\.de via RT wrote&#8203;:
>Thanks for the review\. I have pushed a new version of my branch\.

The entry in embed\.fnc should be flagged "M" to match the "x" on the
apidoc segment\.

\-zefram

**Comment by p5pRT** [comment]
### From @xsawyerx



On 11/10/2017 03&#8203;:11 AM\, l\.mai@&#8203;web\.de via RT wrote&#8203;:
> On Thu\, 09 Nov 2017 16&#8203;:05&#8203;:54 \-0800\, zefram@&#8203;fysh\.org wrote&#8203;:
>> I agree that a wrap\_keyword\_plugin\(\) function would be sensible\.
>> Note that PL\_keyword\_plugin is flagged as experimental\, so
>> wrap\_keyword\_plugin\(\) should also be so marked\, being part of the same
>> experiment\.  \(I have things to say about where that experiment should go\,
>> but that's for another time\.\)  With regard to which mutex to use on older
>> Perls\, I suggest using the OP\_REFCNT mutex again\.  OP\_CHECK\_MUTEX would
>> be another option where it exists\, but that's only back to 5\.15\.8\, so
>> would require reserve definitions to be more complex\, with conditional
>> code to switch to a different mutex on pre\-5\.15\.8 perls\.  OP\_REFCNT is
>> available all the way back\.  There's no real performance concern affecting
>> the choice between these mutexes\, just the maintainability issue\.
>
> In light of this\, my back\-compat suggestion becomes&#8203;:
>
> \#ifndef wrap\_keyword\_plugin
> \#define wrap\_keyword\_plugin\(a\, b\) S\_wrap\_keyword\_plugin\(aTHX\_ a\, b\)
> static void S\_wrap\_keyword\_plugin\(pTHX\_ Perl\_keyword\_plugin\_t new\_plugin\, Perl\_keyword\_plugin\_t \*old\_plugin\_p\) \{
>     dVAR;
>     PERL\_UNUSED\_CONTEXT;
>     if \(\*old\_plugin\_p\) return;
>     MUTEX\_LOCK\(&PL\_op\_mutex\);
>     if \(\!\*old\_plugin\_p\) \{
>         \*old\_plugin\_p = PL\_keyword\_plugin;
>         PL\_keyword\_plugin = new\_plugin;
>     \}
>     MUTEX\_UNLOCK\(&PL\_op\_mutex\);
> \}
> \#endif

Thank you for this patch\, Lukas\!
\(And well\, ya know\, the original work too :\)

Zefram\, while you've reviewed the original patch\, can you also review
this back\-compat patch?

I'll look into merging it in Devel&#8203;::PPPort\.

Thanks\, guys\!

**Comment by p5pRT** [comment]
### From @leonerd

On Wed\, 08 Nov 2017 16&#8203;:34&#8203;:25 \-0800
"l\.mai@&#8203;web\.de via RT" \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> If no one screams\, I'll merge this in a couple of days \(if I
> remember\)\.

A thought occurs to me on this\.

I have two keyword\-plugin\-using modules on CPAN; they are

&nbsp;  Syntax&#8203;::Keyword&#8203;::Try
&nbsp;  Future&#8203;::AsyncAwait

Of those two\, the first one can be made safe using this function\, but
the second I believe still doesn't\, because it also uses a custom op\.

The code currently looks like

&nbsp;  BOOT&#8203;:
&nbsp;    XopENTRY\_set\(&xop\_leaveasync\, xop\_name\, "leaveasync"\);
&nbsp;    XopENTRY\_set\(&xop\_leaveasync\, xop\_desc\, "leaveasync\(\)"\);
&nbsp;    XopENTRY\_set\(&xop\_leaveasync\, xop\_class\, OA\_UNOP\);
&nbsp;    Perl\_custom\_op\_register\(aTHX\_ &pp\_leaveasync\, &xop\_leaveasync\);

&nbsp;    XopENTRY\_set\(&xop\_await\, xop\_name\, "await"\);
&nbsp;    XopENTRY\_set\(&xop\_await\, xop\_desc\, "await\(\)"\);
&nbsp;    XopENTRY\_set\(&xop\_await\, xop\_class\, OA\_UNOP\);
&nbsp;    Perl\_custom\_op\_register\(aTHX\_ &pp\_await\, &xop\_await\);

&nbsp;    next\_keyword\_plugin = PL\_keyword\_plugin;
&nbsp;    PL\_keyword\_plugin = &my\_keyword\_plugin;

I would need a suitable locking mechanism to lock the entire thing\.
Your suggested wrapper function for setting the PL\_keyword\_plugin hook
would only protect these final two lines; I shall need something
further for locking the XOP\-setting code\.


I begin to wonder if this is a wider problem with BOOT itself; that
what is needed is a way to ensure it is executed exactly once when when
late\-loaded in a multi\-threaded environment\. Maybe that can be
constructed out of a suitable macro thus&#8203;:

&nbsp;  static boot\_once\(pTHX\) \{ \.\.\. \}

&nbsp;  BOOT&#8203;:
&nbsp;    ONCE\(&boot\_once\);

\-\- 
Paul "LeoNerd" Evans

leonerd@&#8203;leonerd\.org\.uk      |  https://metacpan.org/author/PEVANS
http&#8203;://www\.leonerd\.org\.uk/  |  https://www.tindie.com/stores/leonerd/

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Sawyer X wrote&#8203;:
>Zefram\, while you've reviewed the original patch\, can you also review
>this back\-compat patch?

The proposed reserve definition of wrap\_keyword\_plugin\(\) is mostly
correct\, but should use the OP\_REFCNT\_\{\,UN\}LOCK macros rather than
MUTEX\_\{\,UN\}LOCK\(\)\, to be portable to unthreaded builds\.

\-zefram

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Paul "LeoNerd" Evans wrote&#8203;:
>I begin to wonder if this is a wider problem with BOOT itself; that
>what is needed is a way to ensure it is executed exactly once when when
>late\-loaded in a multi\-threaded environment\.

No\, there are parts of BOOT that need to run per interpreter\.  The main
hidden part of an XS boot\, setting up XSUB CVs\, needs to be done for each
interpreter in which the module is loaded\, and any CV creation that's
done in an explicit BOOT section needs to be treated that way too\.
The things that are shared between interpreters are more the exceptions\.

\-zefram

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun\, 12 Nov 2017 16&#8203;:20&#8203;:47 \-0800\, zefram@&#8203;fysh\.org wrote&#8203;:
> Paul "LeoNerd" Evans wrote&#8203;:
> >I begin to wonder if this is a wider problem with BOOT itself; that
> >what is needed is a way to ensure it is executed exactly once when when
> >late\-loaded in a multi\-threaded environment\.
> 
> No\, there are parts of BOOT that need to run per interpreter\.  The main
> hidden part of an XS boot\, setting up XSUB CVs\, needs to be done for each
> interpreter in which the module is loaded\, and any CV creation that's
> done in an explicit BOOT section needs to be treated that way too\.
> The things that are shared between interpreters are more the exceptions\.
> 
> \-zefram

Do you think a BOOTONCE section would be an appropriate extension to XS\, or is it rare enough that we do not need it?

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @mauke

Am 13\.11\.2017 um 01&#8203;:16 schrieb Zefram&#8203;:
> Sawyer X wrote&#8203;:
>> Zefram\, while you've reviewed the original patch\, can you also review
>> this back\-compat patch?
> 
> The proposed reserve definition of wrap\_keyword\_plugin\(\) is mostly
> correct\, but should use the OP\_REFCNT\_\{\,UN\}LOCK macros rather than
> MUTEX\_\{\,UN\}LOCK\(\)\, to be portable to unthreaded builds\.

MUTEX\_LOCK/MUTEX\_UNLOCK are NOOPs on unthreaded builds\.

\-\- 
Lukas Mai \<plokinom@&#8203;gmail\.com>

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Father Chrysostomos via RT wrote&#8203;:
>Do you think a BOOTONCE section would be an appropriate extension to XS\,
>or is it rare enough that we do not need it?

It sounds risky\, because of the greatly expanded scope of the critical
section\, given the need to mutex it\.  I think we should continue to
add purpose\-specific idempotence and mutexing where it's needed\.  For a
module to require such treatment for a data structure of its own is rare
enough to be not worth a new XS section\.

\-zefram

**Comment by p5pRT** [comment]
### From @mauke

On Wed\, 08 Nov 2017 16&#8203;:34&#8203;:24 \-0800\, mauke\- wrote&#8203;:
> On Tue\, 07 Nov 2017 09&#8203;:44&#8203;:49 \-0800\, mauke\- wrote&#8203;:
> >
> > PL\_check \(the other global variable documented in perlapi\) has this
> > paragraph
> > in its documentation&#8203;:
> >
> > | For thread safety\, modules should not write directly to this array\.
> > Instead\,
> > | use the function wrap\_op\_checker\.
> >
> > This equally applies to PL\_keyword\_plugin\, but a corresponding
> > wrap\_keyword\_plugin function is missing\. It should be added\.
> 
> I have done this in a branch&#8203;:
> https://perl5.git.perl.org/perl.git/shortlog/refs/heads/smoke-
> me/mauke/keyword\-plugin\-mutex
> 
> If no one screams\, I'll merge this in a couple of days \(if I
> remember\)\.

The branch has been merged in commit 6cc7638e57c54706dc2d698d9b2f9f769c17ffb4\. Should the status of this ticket change to "pending release"?

**Comment by p5pRT** [comment]
### From @xsawyerx



On 11/13/2017 08&#8203;:25 PM\, l\.mai@&#8203;web\.de via RT wrote&#8203;:
> On Wed\, 08 Nov 2017 16&#8203;:34&#8203;:24 \-0800\, mauke\- wrote&#8203;:
>> On Tue\, 07 Nov 2017 09&#8203;:44&#8203;:49 \-0800\, mauke\- wrote&#8203;:
>>> PL\_check \(the other global variable documented in perlapi\) has this
>>> paragraph
>>> in its documentation&#8203;:
>>>
>>> | For thread safety\, modules should not write directly to this array\.
>>> Instead\,
>>> | use the function wrap\_op\_checker\.
>>>
>>> This equally applies to PL\_keyword\_plugin\, but a corresponding
>>> wrap\_keyword\_plugin function is missing\. It should be added\.
>> I have done this in a branch&#8203;:
>> https://perl5.git.perl.org/perl.git/shortlog/refs/heads/smoke-
>> me/mauke/keyword\-plugin\-mutex
>>
>> If no one screams\, I'll merge this in a couple of days \(if I
>> remember\)\.
> The branch has been merged in commit 6cc7638e57c54706dc2d698d9b2f9f769c17ffb4\. Should the status of this ticket change to "pending release"?

Yes\, but I would like to merge the backport solution to Devel&#8203;::PPPort\.

Zefram\, can you respond on Lukas' latest comment on
MUTEX\_LOCK/MUTEX\_UNLOCK being a NOOP on unthreaded builds? Does it mean
we can copy the latest version Lukas had?

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Sawyer X wrote&#8203;:
>Zefram\, can you respond on Lukas' latest comment on
>MUTEX\_LOCK/MUTEX\_UNLOCK being a NOOP on unthreaded builds?

He's right\.  His version is fine\.

\-zefram
